documentation/ADRs/014_input_drift_detection.md

```diff
@@ -0,0 +1,38 @@
+# Drift-detection
+
+| ADR Info          | Details               |
+|-------------------|-----------------------|
+| Subject           | Input drift detection |
+| ADR Number        | 014                   |
+| Status            | proposed              |
+| Author            | Jim Dale              |
+| Date              | 02/10/2024            |
+
+## Context
+The pipeline needs machinery for checking that the input data entering it has not been compromised by problems or errors in the monthly data ingestion. This machinery also needs the ability to test itself.
+
+## Decision
+In discussions involving Simon Polichinel von der Maase, Mihai Croicu and James Dale at the PRIO cabin retreat in Feb 2024, it was decided that input drift detection should be done by the viewser data-fetching client itself, so that drift-detection can be conducted at any time when data is read into the pipeline.
+
+## Consequences
+
+**Positive Effects:**
```

Comment:
### 1. **Data Ingestion Failure Monitoring**:
   - **Consistency & Integrity**: Ensures data is received consistently and without corruption. Detects missing, delayed, or faulty data early.
   - **Alerts for Pipeline Issues**: Provides early warning if ingestion fails, allowing for quick intervention and preventing faulty predictions.
   - **Fresh Data Verification**: Ensures models operate on up-to-date data through timestamp checks, preventing stale data from affecting predictions.

### 2. **Detecting Large Shifts in Input Data**:
   - **Real-World Event Alerts**: Identifies significant shifts in data that may indicate important real-world changes, such as spikes in conflict.
   - **Performance Context**: Helps explain performance drops by identifying whether changes in data reflect actual developments or pipeline issues.

### 3. **Monitoring Data Source Reliability**:
   - **Detecting 'Bad' Data**: Alerts when data sources degrade in quality or become unreliable, preventing contamination of model predictions.
   - **Dynamic Validation**: Compares new data against historical patterns, flagging anomalies or corrupted inputs.

### 4. **Impact on Model & Pipeline**:

   - **Integrated Alerts**: Combines data and model monitoring to provide clear context for performance changes - whether due to real-world shifts or technical issues.
   - **Adaptive Systems**: Triggers automatic model retraining when significant input drift or data problems are detected.

Reply:
Most of this I agree with and is nice :) A few issues though:

(i) I'm not sure what you mean by 'delayed data' in section 1 above?

(ii) Fresh Data Verification - the input drift-detection system does not do any checking of timestamps - is this something which the dataloaders themselves should do?

(iii) in section 4, 'combines data and model monitoring' - the  input drift system does not know anything about models - in combination with the *output* drift system (when we have one ;) ) it could do this, but it does not do it on its own. 

(iv) Also in section 4, adaptive systems - this is functionality which we would like to have but do not have yet

Reply:
So this was just examples, but quick answers:

i) I just meant data that for some reason had not been updated this month. But for our purpose  that is probability just the same as missing 

ii) Hmm... yeah, no. This is not strictly drift detection. And we are moving towards doing it in the dataloaders - at least for the UCDP. But there should be a thing somewhere to check everything.  Anyway don't worry about it now. 

iii) Again, example. Don't worry about it

iv) Sure - don't worry about it